udffsck/udffsck.c

```diff
@@ -0,0 +1,3093 @@
+
+    struct impUseLVID *impUse = (struct impUseLVID *)((uint8_t *)(disc->udf_lvid) + sizeof(struct logicalVolIntegrityDesc) + 8*disc->udf_lvid->numOfPartitions); //this is because of ECMA 167r3, 3/24, fig 22
+    uint8_t *impUseArr = (uint8_t *)impUse;
+    stats->actUUID = (((struct logicalVolHeaderDesc *)(disc->udf_lvid->logicalVolContentsUse))->uniqueID);
+
+    stats->LVIDtimestamp = lvid->recordingDateAndTime;
+
+    dbg("LVID: number of files: %d\n", impUse->numOfFiles);
+    dbg("LVID: number of dirs:  %d\n", impUse->numOfDirs);
+    dbg("LVID: UDF rev: min read:  %04x\n", impUse->minUDFReadRev);
+    dbg("               min write: %04x\n", impUse->minUDFWriteRev);
+    dbg("               max write: %04x\n", impUse->maxUDFWriteRev);
+    dbg("Next Unique ID: %d\n", stats->actUUID); 
+    dbg("LVID recording timestamp: %s\n", print_timestamp(stats->LVIDtimestamp)); 
+
+    stats->expNumOfFiles = impUse->numOfFiles;
+    stats->expNumOfDirs = impUse->numOfDirs;
+
+    stats->minUDFReadRev = impUse->minUDFReadRev;
+    stats->minUDFWriteRev = impUse->minUDFWriteRev;
+    stats->maxUDFWriteRev = impUse->maxUDFWriteRev;
+
+    dbg("Logical Volume Contents Use\n");
+    for(int i=0; i<32; ) {
+        for(int j=0; j<8; j++, i++) {
+            note("%02x ", disc->udf_lvid->logicalVolContentsUse[i]);
+        }
+        note("\n");
+    }
+    dbg("Free Space Table\n");
+    for(int i=0; i<disc->udf_lvid->numOfPartitions * 4; i++) {
+        note("0x%08x, %d\n", disc->udf_lvid->freeSpaceTable[i], disc->udf_lvid->freeSpaceTable[i]);
+    }
+    stats->freeSpaceBlocks = disc->udf_lvid->freeSpaceTable[0];
+    stats->partitionSizeBlocks = disc->udf_lvid->freeSpaceTable[1];
+
+    dbg("Size Table\n");
+    for(int i=disc->udf_lvid->numOfPartitions * 4; i<disc->udf_lvid->numOfPartitions * 4 * 2; i++) {
+        note("0x%08x, %d\n", disc->udf_lvid->freeSpaceTable[i],disc->udf_lvid->freeSpaceTable[i]);
+    }
+
+    if(disc->udf_lvid->nextIntegrityExt.extLength > 0) {
+        dbg("Next integrity extent found.\n");
+    } else {
+        dbg("No other integrity extents are here.\n");
+    }
+
+    unmap_chunk(dev, chunk, st_size); 
+    return 0; 
+}
+
+/**
+ * \brief Select various volume identifiers and store them at stats structure
+ * 
+ * At this moment it selects PVD->volSetIdent and FSD->logicalVolIdent
+ *
+ * \param[in] *disc disc structure
+ * \param[out] *stats file system status structure
+ * \param[in] *seq VDS sequence
+ *
+ * \return 0 -- everything OK
+ * \return 4 -- no correct PVD found.
+ */
+int get_volume_identifier(struct udf_disc *disc, struct filesystemStats *stats, vds_sequence_t *seq ) {
+    int vds = -1;
+    if((vds=get_correct(seq, TAG_IDENT_PVD)) < 0) {
+        err("No correct PVD found. Aborting.\n");
+        return 4;
+    }
+    char *namebuf = calloc(1,128*2);
+    memset(namebuf, 0, 128*2);
+    int size = decode_utf8(disc->udf_pvd[vds]->volSetIdent, namebuf, 128);
```

Comment:
Function `gen_uuid_from_vol_set_ident()` is now in master branch https://github.com/pali/udftools/commit/9d60f90ea6c88b52aad0d18ff33f49605e27b2e9